### PR TITLE
Fix WCOS/Win32 linking bugs

### PR DESCRIPTION
**Pull request by tiagoshibata**
Edit: Scope of this PR has changed. After some discussion it seems like ORT will build for WCOS after version 1.2. I'm removing the WCOS/Win32 hacks and using the CMake cross-compilation options to target WCOS or Win32. When using the VS generator, this also sets the project type in VS, so it's easy to see in the project pane if you're targeting Win32 or Store.

 ----

**Description**: WCOS/Win32 linking options were error prone, because CMAKE_CXX_STANDARD_LIBRARIES was left at the default value (which includes Win32 libs). The WCOS targets were being linked with both Win32 and WCOS libs, which links with the wrong library if you swap the order of the libraries (e.g. refactoring), and can lead to subtle bugs by accidentally calling a Win32 function and linking to a Win32 library.

### CODE REVIEW

File: cmake/winml.cmake [hunk 1]
```diff
@@ -604,4 +604,4 @@ endif()
 target_link_options(winml_dll PRIVATE /ignore:4199)
 
 target_link_libraries(winml_dll PRIVATE windowsapp.lib)
-target_link_options(winml_dll PRIVATE /NODEFAULTLIB:kernel32.lib /NODEFAULTLIB:user32.lib /NODEFAULTLIB:gdi32.lib /NODEFAULTLIB:winspool.lib /NODEFAULTLIB:shell32.lib /NODEFAULTLIB:ole32.lib /NODEFAULTLIB:oleaut32.lib /NODEFAULTLIB:uuid.lib /NODEFAULTLIB:comdlg32.lib /NODEFAULTLIB:advapi32.lib)
\ No newline at end of file
+target_link_options(winml_dll PRIVATE /NODEFAULTLIB:kernel32.lib /NODEFAULTLIB:user32.lib /NODEFAULTLIB:gdi32.lib /NODEFAULTLIB:winspool.lib /NODEFAULTLIB:shell32.lib /NODEFAULTLIB:ole32.lib /NODEFAULTLIB:oleaut32.lib /NODEFAULTLIB:uuid.lib /NODEFAULTLIB:comdlg32.lib /NODEFAULTLIB:advapi32.lib)
```
Comment (orilevari) [inline cmake/winml.cmake]: This is no longer necessary with windows_toolchain.cmake correct?

File: cmake/windows_toolchain.cmake [hunk 2]
```diff
@@ -0,0 +1,38 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+# The default libraries to link with in Windows are Win32 libs:
+# kernel32.lib;user32.lib;gdi32.lib;winspool.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;comdlg32.lib;advapi32.lib
+# If we're targeting WCOS, we must override them with a WCOS umbrella library.
+# Else, onnxruntime must use the Win32 libs to support downlevel platforms, and WinML must use the umbrella library.
+# CMake doesn't support multiple toolchains in the same project, so in this case we use a small hack clearing
```
Comment (smk2007) [inline cmake/windows_toolchain.cmake]: >small hack clearing [](start = 90, length = 19)

dont think we need the hack if everything is built using wcos, right?

File: tools/ci_build/build.py [hunk 3]
```diff
@@ -165,7 +165,7 @@ def parse_arguments():
     parser.add_argument("--use_dml", action='store_true', help="Build with DirectML.")
     parser.add_argument("--use_winml", action='store_true', help="Build with WinML.")
     parser.add_argument("--use_telemetry", action='store_true', help="Only official builds can set this flag to enable telemetry.")
-    parser.add_argument("--enable_wcos", action='store_true', help="Build for Windows Core OS.")
+    parser.add_argument("--win32", action='store_true', help="Build for Win32 instead of Windows Core OS APIs")
```
Comment (smk2007) [inline tools/ci_build/build.py]: we had agreed a little earlier on that the default behavior for the build should be win32, as there are other pipeline build (like for edge) that require a win32 and are not explicitly configured for win32.

File: cmake/CMakeLists.txt [hunk 4]
```diff
@@ -4,16 +4,19 @@
...
 # Minimum CMake required
 cmake_minimum_required(VERSION 3.13)
-
-# Project
-project(onnxruntime C CXX)
 cmake_policy(SET CMP0069 NEW)
-set(CMAKE_POLICY_DEFAULT_CMP0069 NEW)
-
 if (${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.15")
+  if (NOT onnxruntime_USE_CUDA)
```
Comment (smk2007) [inline cmake/CMakeLists.txt]: why was this change needed?

File: cmake/CMakeSettings.json [hunk 5]
```diff
@@ -0,0 +1,62 @@
+{
+  "configurations": [
+    {
+      "name": "WinML-Debug",
+      "generator": "Ninja",
```
Comment (smk2007) [inline cmake/CMakeSettings.json]: Is "Ninja" intentional here?

File: cmake/CMakeSettings.json [hunk 6]
```diff
@@ -0,0 +1,62 @@
+{
+  "configurations": [
+    {
+      "name": "WinML-Debug",
```
Comment (smk2007) [inline cmake/CMakeSettings.json]: These configurations are manually set with build.py right? Why do we need these?

File: cmake/CMakeSettings.json (hunk 5 above)
Comment (tiagoshibata) [reply cmake/CMakeSettings.json]: Yes, that's the default generated by VS

File: cmake/CMakeSettings.json (hunk 6 above)
Comment (tiagoshibata) [reply cmake/CMakeSettings.json]: Native VS integration with CMake

File: cmake/CMakeLists.txt [hunk 7]
```diff
@@ -122,6 +124,11 @@ else()
     set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /d2FH4-")
     set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /d2FH4-")
   endif()
+  if(WINDOWS_STORE)
+    # cmake/external/protobuf/src/google/protobuf/compiler/subprocess.cc and onnxruntime/core/platform/windows/env.cc call a bunch of Win32 APIs.
```
Comment (snnn) [inline cmake/CMakeLists.txt]: @tiagoshibata , could you please give us a guildline of which APIs can be used, which are not?

CC: @edgchen1

Comment (tiagoshibata) [reply cmake/CMakeLists.txt]: Compiling with WINAPI_FAMILY set to WINAPI_FAMILY_APP (the default in UWP mode) will make headers hide Win32-only definitions. That's an easy way to find all the bad API calls.

I don't know if there's a definitive list somewhere... Opening the API docs under https://docs.microsoft.com/en-us/windows/win32/api shows on the bottom of the page if it's Desktop only, e.g. GetFileSizeEx:

2020-03-17-10:24:57-screenshot

Versus LoadLibraryA:

2020-03-17-10:25:09-screenshot

File: cmake/CMakeLists.txt [hunk 8]
```diff
@@ -122,6 +119,11 @@ else()
...
+    # For now, we'll set the API family to desktop globally to expose Win32 symbols in headers; this must be fixed!
```
Comment (martinb35) [inline cmake/CMakeLists.txt]: is there a task tracking the "this must be fixed!" part?

File: cmake/CMakeSettings.json [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        {
+          "name": "onnxruntime_BUILD_SHARED_LIB",
+          "value": "True",
+          "type": "BOOL"
+        }
+      ],
+      "cmakeToolchain": "${projectDir}\\wcos_toolchain.cmake"
+    }
+  ]
+}
```
Comment (martinb35) [inline cmake/CMakeSettings.json]: is the settings file intentional? looks like it might be user-specific

File: cmake/CMakeLists.txt (hunk 8 above)
Comment (tiagoshibata) [reply cmake/CMakeLists.txt]: This is on the ORT side. Changming is aware of it: https://github.com/microsoft/onnxruntime/pull/3126#discussion_r393834602

File: cmake/CMakeSettings.json (hunk 9 above)
Comment (tiagoshibata) [reply cmake/CMakeSettings.json]: Intentional. It sets default configurations to show when you open the CMake project in VS (like normal VS projects have Debug/Release and platform choices, when using the VS CMake integration you can set default configs in the CMakeSettings.json file). Here I added a couple of configs for WCOS.

I'm going to show to the team the workflow using native VS integration vs CMake command line vs build.py and we can discuss what works best.

File: cmake/CMakeLists.txt (hunk 7 above)
Comment (tiagoshibata) [reply cmake/CMakeLists.txt]: https://docs.microsoft.com/en-us/windows/win32/apiindex/umbrella-lib-onecoreuap
